### PR TITLE
Implement a fitsort mode for `fitsheader`.

### LINKED ISSUE #7055: fitsheader as an alternative to dfits+fitsort ?
At the moment if I run `fitsheader -t -e 0 -k "KEYWORD1" -k "KEYWORD2" *.fits`, I obtain something that looks like this:

    |   filename | hdu |  keyword | value |
    | file1.fits |  0  | KEYWORD1 |     1 |
    | file1.fits |  0  | KEYWORD2 |     2 |
    | file2.fits |  0  | KEYWORD1 |     1 |
    | file2.fits |  0  | KEYWORD2 |     3 |
    | file3.fits |  0  | KEYWORD1 |     1 |
    | file3.fits |  0  | KEYWORD2 |     4 |

Which is not very convenient to compare the values of some keywords across multiple files.
I would like to find a way to get the output to look like the following, which is the behavior of the good old `fits` and `fitsort` tools:

    |   filename | KEYWORD1 | KEYWORD2 |
    | file1.fits |        1 |        2 |
    | file2.fits |        1 |        3 |
    | file3.fits |        1 |        4 |

If the hdu index of the keywords matters, the format could be tweaked to the following:

    |   filename | 0:KEYWORD1 | 0:KEYWORD2 |
    | file1.fits |          1 |          2 |
    | file2.fits |          1 |          3 |
    | file3.fits |          1 |          4 |

I am happy to implement this (assuming that something equivalent is not already available), but wanted to have some feedback first.

### PR DESCRIPTION
**Pull request by jwoillez**
Fixes #7055.

- [x] Update the documentation.
- [x] Implement dotted keyword handling.

### CODE REVIEW

**Comment by jwoillez** [comment]
This should be good to go, including the dotted notation for keywords.

**Comment by saimn** [comment]
Looks great, thanks @jwoillez ! I will look (and test) more in detail later, could you just add a unit test ? (in `astropy/io/fits/tests/test_fitsheader.py`)

**Comment by jwoillez** [comment]
@saimn I added two tests for the `fitsort` mode, with and without sorting. I will add a test for the keyword dotted notation a bit later.

**Comment by jwoillez** [comment]
@saimn The test for the dotted keyword notation is also there.

File: astropy/io/fits/scripts/fitsheader.py [hunk 1]
```diff
@@ -275,9 +292,76 @@ def print_headers_as_table(args):
...
+            keyword = values.replace('.',' ')
+            if namespace.keywords is None:
+                namespace.keywords = []
+            if keyword not in namespace.keywords:
+                namespace.keywords.append(keyword)
+            if option_string in ['-f','--fitsort']:
+                namespace.fitsort = keyword
+        else:
+            if option_string in ['-f','--fitsort']:
+                namespace.fitsort = True
```
Comment (saimn) [inline astropy/io/fits/scripts/fitsheader.py]: I don't understand why `fitsort` is managed here. It gives weird result if I use `fitsheader --fitsort astropy/io/fits/tests/data/test1.fits` which seems valid (or should give an error if specifying keywords is required).

File: astropy/io/fits/scripts/fitsheader.py [hunk 2]
```diff
@@ -309,16 +400,18 @@ def main(args=None):
...
                              'wildcards are supported')
     args = parser.parse_args(args)
-
+    
     # If `--table` was used but no format specified,
     # then use ascii.fixed_width by default
     if args.table is None:
         args.table = 'ascii.fixed_width'
-
+
```
Comment (saimn) [inline astropy/io/fits/scripts/fitsheader.py]: These extra whitespaces should be removed.

File: astropy/io/fits/scripts/fitsheader.py [hunk 3]
```diff
@@ -275,9 +292,76 @@ def print_headers_as_table(args):
...
+        return False
+    elif len(tables) == 1:
+        resulting_table = tables[0]
+    else:
+        from .... import table
+        resulting_table = table.vstack(tables)
+    # If we obtained more than one hdu, merge hdu and keywords columns
+    if len(np.unique(resulting_table['hdu']))>1:
+        for tab in tables:
+            new_column = table.Column(['{}:{}'.format(row['hdu'],row['keyword']) for row in tab])
```
Comment (saimn) [inline astropy/io/fits/scripts/fitsheader.py]: `table` is not defined if printing keywords for multiple extensions with only one file (`fitsheader --fitsort -k EXPSTART -k EXPTIME -k ESO.DET.ID astropy/io/fits/tests/data/test1.fits`).

File: astropy/io/fits/scripts/fitsheader.py [hunk 4]
```diff
@@ -275,9 +292,76 @@ def print_headers_as_table(args):
...
+            tab.add_column(new_column, name='hdu+keyword')
+        keyword_column_name = 'hdu+keyword'
+    else:
+        keyword_column_name = 'keyword'
+    # Check how many hdus we are processing
+    final_tables = []
+    for tab in tables:
+        final_table = [table.Column([tab['filename'][0]], name='filename')]
```
Comment (saimn) [inline astropy/io/fits/scripts/fitsheader.py]: Same here, `table` is not defined with one file.

**Comment by saimn** [comment]
The `Keyword XXX not found.` warnings are annoying when one wants to print many keywords from multiple files which may not have all keywords, so I was thinking that we could add a quiet (`-q`, `--quiet`) option to shut off these warnings. Thoughts ?

**Comment by jwoillez** [comment]
@saimn Thanks for the feedback. Never thought of testing `--fitsort` on less than 2 files. Here is a new version hopefully including fixes for all the issues you reported. With respect to the `Keyword XXX not found` messages, how about disabling them by default in `--fitsort` mode?

**Comment by saimn** [comment]
> With respect to the Keyword XXX not found messages, how about disabling them by default in --fitsort mode?

Yes, I think it makes sense since a typical usecase of fitsort is to parse keywords for many files with possibly missing keywords.

File: astropy/io/fits/scripts/fitsheader.py [hunk 5]
```diff
@@ -275,9 +292,71 @@ def print_headers_as_table(args):
...
+    from .... import table
+    tables = []
+    # Create a Table object for each file
+    for filename in args.filename:  # Support wildcards
+        try:
+            formatter = TableHeaderFormatter(filename)
+            tbl = formatter.parse(args.extensions,
+                                  args.keywords,
+                                  args.compressed)
+            tbl = tbl[np.where(tbl['keyword']!='')]  # Remove empty keywords
```
Comment (saimn) [inline astropy/io/fits/scripts/fitsheader.py]: `np.where` is probably useless here (I like to track the *Useless Use of np.where* :smile: ).

File: astropy/io/fits/scripts/fitsheader.py [hunk 6]
```diff
@@ -275,9 +292,71 @@ def print_headers_as_table(args):
...
+        final_table = [table.Column([tab['filename'][0]], name='filename')]
+        for row in tab:
+            final_table.append(table.Column([row['value']], name=row[keyword_column_name]))
+        final_tables.append(table.Table(final_table))
+    final_table = table.vstack(final_tables)
+    # Sort if requested
+    if args.fitsort is not True:  # then it must be a keyword, therefore sort
+        final_table.sort(args.fitsort)
+    # Reorganise to keyword by columns
+    final_table.pprint(max_lines=-1, max_width=-1, show_dtype=True)
```
Comment (saimn) [inline astropy/io/fits/scripts/fitsheader.py]: Is it useful to show the dtype ? IMO it isn't, I don't think it brings a useful information.

File: astropy/io/fits/tests/test_fitsheader.py [hunk 7]
```diff
@@ -77,3 +77,34 @@ def test_table(self, capsys):
...
+        out = out.splitlines()
+        assert len(out) == 2
+        assert out[1].strip().endswith("HIERARCH ESO DET ID = 'DV13' / Detector system Id")
+
+        fitsheader.main(['-e', '0', '-k', 'ESO.DET.ID',
+                         self.data('fixed-1890.fits')])
+        out, err = capsys.readouterr()
+        out = out.splitlines()
+        assert len(out) == 2
+        assert out[1].strip().endswith("HIERARCH ESO DET ID = 'DV13' / Detector system Id")
```
Comment (saimn) [inline astropy/io/fits/tests/test_fitsheader.py]: Could you also add a test for the table header line with multiple HDUs ? E.g. testing something like this:
```
❯ fitsheader -f -k NAXIS astropy/io/fits/tests/data/t*
               filename               0:NAXIS 1:NAXIS 2:NAXIS 3:NAXIS 4:NAXIS
                str37                   str1    str1    str1    str1    str1 
------------------------------------- ------- ------- ------- ------- -------
astropy/io/fits/tests/data/table.fits       0       2      --      --      --
   astropy/io/fits/tests/data/tb.fits       0       2      --      --      --
 astropy/io/fits/tests/data/tdim.fits       0       2      --      --      --
astropy/io/fits/tests/data/test0.fits       0       2       2       2       2
astropy/io/fits/tests/data/test1.fits       0       2       2       2       2
```

File: astropy/io/fits/scripts/fitsheader.py (hunk 5 above)
Comment (jwoillez) [reply astropy/io/fits/scripts/fitsheader.py]: If you run `fitsheader astropy/io/fits/tests/data/test0.fits`, you will see keyword-less lines. We need to get rid of these. They could even cause a crash due to duplicated columns (two empty keywords in the same fits header).

File: astropy/io/fits/scripts/fitsheader.py (hunk 6 above)
Comment (jwoillez) [reply astropy/io/fits/scripts/fitsheader.py]: Should I coud replace it by showing the units?

File: astropy/io/fits/tests/test_fitsheader.py (hunk 7 above)
Comment (jwoillez) [reply astropy/io/fits/tests/test_fitsheader.py]: Done!

**Comment by jwoillez** [comment]
Note: I am still getting a duplicate columns errors with `fitsheader -f t*.fits`, and have to figure out why. Does the fits standard allow to have the same keyword twice in the same HDU header?

File: astropy/io/fits/scripts/fitsheader.py (hunk 5 above)
Comment (saimn) [reply astropy/io/fits/scripts/fitsheader.py]: Yes sure, sorry for the confusion, I meant that `np.where` is useless, not the filtering. You can use `tbl[tbl['keyword'] != '']`.

File: astropy/io/fits/scripts/fitsheader.py (hunk 6 above)
Comment (saimn) [reply astropy/io/fits/scripts/fitsheader.py]: FITS keywords don't have units ?

**Comment by saimn** [comment]
> Does the fits standard allow to have the same keyword twice in the same HDU header?

Yes ! In your example the issue is with COMMENT I think, but it could be for HISTORY and actually any keyword can be duplicated.

Could you rebase on master to get the unit tests fixes ?

File: astropy/io/fits/scripts/fitsheader.py (hunk 6 above)
Comment (jwoillez) [reply astropy/io/fits/scripts/fitsheader.py]: Indeed... confused by astropy's `Table`. Will remove `dtype`.

**Comment by jwoillez** [comment]
How would you deal with duplicate keywords (beside HISTORY and COMMENT) in a same HDU? Post-fix an index?

**Comment by saimn** [comment]
@jwoillez - Yes, post-fixing seems reasonable.

**Comment by saimn** [comment]
Hi @jwoillez ,
It would be great to finish this, do you think you can find some time ? I can have a look otherwise. I think we are just missing the duplicate keywords handling?

**Comment by saimn** [comment]
Also it needs a rebase to fix the changelog conflict.

**Comment by jwoillez** [comment]
I did not have much time to work this.
The FITS standard states the following:

> The mandatory FITS keywords defined in this standard must not appear more than once within a header. All other keywords that have a value should not appear more than once. If a keyword does appear multiple times with different values, then the value is indeterminate.

Would taking the last value for repeating keywords and ignoring `COMMENT` and `HISTORY` be OK?

**Comment by saimn** [comment]
@jwoillez - Ah great, I missed this paragraph (and was confused by the fact that `fits.Header` can manage duplicate keywords). So yes, you can just ignore COMMENT and HISTORY, and it will good to go :)

**Comment by saimn** [comment]
@jwoillez - I rebased the branch and added a few commits, to exclude COMMENT and HISTORY, and to add files in the output even if they don't have the required keyword, e.g.:
```
 fitsheader -f astropy/io/fits/tests/data/t* -k DATE-OBS -k ORIGIN
               filename                               ORIGIN                 DATE-OBS
------------------------------------- -------------------------------------- --------
astropy/io/fits/tests/data/table.fits                                     --       --
   astropy/io/fits/tests/data/tb.fits                    STScI-STSDAS/TABLES       --
 astropy/io/fits/tests/data/tdim.fits                                     --       --
astropy/io/fits/tests/data/test0.fits NOAO-IRAF FITS Image Kernel Aug 1 1997 19/05/94
astropy/io/fits/tests/data/test1.fits NOAO-IRAF FITS Image Kernel Aug 1 1997 19/05/94
```
If you want to have a final look, is it good for you ?

**Comment by saimn** [comment]
I'm going to merge this as I think that it's in good shape, and it would be good to have enough time to test this on real cases before the next release. Thanks @jwoillez !

**Comment by jwoillez** [comment]
@saimn Thank you for taking over! I somehow did not manage to find the time to finish this.

**Pull request merged**